### PR TITLE
fix: wait until playground is initialized

### LINKED ISSUE #147: Detect the playground container is ready
The playground container can get some time to start, and the user can send a request before the container is ready to reply.

It would be helpful to implement a readiness probe, so AI studio knows when the playground container is ready to reply.

### PR DESCRIPTION
**Pull request by jeffmaury**
Fixes #147

### What does this PR do?

Wait until the playground container is properly initialized

### Screenshot / video of UI

N/A

### What issues does this PR fix or reference?

Fixes #147 

### How to test this PR?

Start a playground

### CODE REVIEW

**Review by benoitf — APPROVED** [review]
LGTM

File: packages/backend/src/managers/playground.ts [hunk 1]
```diff
@@ -205,6 +207,26 @@ export class PlayGroundManager {
...
 
+    let contacted = false;
+    const start = Date.now();
+    while (
+      Date.now() - start < STARTING_TIME_MAX &&
+      !contacted &&
+      this.playgrounds.get(modelId).status === 'starting'
+    ) {
+      try {
+        await fetch(`http://localhost:${freePort}`);
```
Comment (dgolovin) [inline packages/backend/src/managers/playground.ts:218]:
```suggestion
        const response  = await fetch(`http://localhost:${freePort}`);
```

File: packages/backend/src/managers/playground.ts [hunk 2]
```diff
@@ -205,6 +207,26 @@ export class PlayGroundManager {
...
+        contacted = true;
```
Comment (dgolovin) [inline packages/backend/src/managers/playground.ts:219]:
```suggestion
        contacted = response.ok;
```

Comment (dgolovin) [reply packages/backend/src/managers/playground.ts:219]: nit: There are some use cases when response has error codes that does not trigger exceptions and host would be still unresponsive (network error in the 4xx–5xx range for example).

**Review by dgolovin — APPROVED** [review]

**Comment by feloy** [comment]
The test are failing with heap memory limit. Is the `while` loop capped by time terminating during the tests?

File: packages/backend/src/managers/playground.ts [hunk 3]
```diff
@@ -205,6 +207,26 @@ export class PlayGroundManager {
...
+    let contacted = false;
+    const start = Date.now();
+    while (
+      Date.now() - start < STARTING_TIME_MAX &&
+      !contacted &&
+      this.playgrounds.get(modelId).status === 'starting'
+    ) {
+      try {
+        const response = await fetch(`http://localhost:${freePort}`);
+        contacted = response.ok;
```
Comment (feloy) [inline packages/backend/src/managers/playground.ts:219]: Some sleep between two tries would be less resource consuming

**Comment by feloy** [comment]
I cannot make it work on mac. Let me add some sleep between requests to be able to check the logs, or I get by logs
```
INFO:     192.168.127.1:56827 - "GET / HTTP/1.1" 404 Not Found
```

**Comment by jeffmaury** [comment]
> The test are failing with heap memory limit. Is the `while` loop capped by time terminating during the tests?

It is and seems related to another (flaky ?) test

File: packages/backend/src/managers/playground.ts (hunk 3 above)
Comment (jeffmaury) [reply packages/backend/src/managers/playground.ts:220]: Added

File: packages/backend/src/managers/playground.ts [hunk 4]
```diff
@@ -217,6 +218,7 @@ export class PlayGroundManager {
       try {
         const response = await fetch(`http://localhost:${freePort}`);
         contacted = response.ok;
```
Comment (feloy) [inline packages/backend/src/managers/playground.ts:220]: On my side (mac), I get a response with value `{}`. 
I would set `contacted = true` here, as the socket connection errors we want to detect go to the catch, so if we are at this point, socket connection is ok

Comment (jeffmaury) [reply packages/backend/src/managers/playground.ts]: Should be fixed by 528e991

**Review by feloy — APPROVED** [review]
works on mac

File: packages/backend/src/managers/playground.ts [hunk 5]
```diff
@@ -204,6 +207,26 @@ export class PlayGroundManager {
       }
     });
 
+    let contacted = false;
```
Comment (axel7083) [inline packages/backend/src/managers/playground.ts:210]: Maybe we should create an utility function _somewhere_ that would manage the `retry` logics, we would reuse it later for other usage, with some timeout limits. See https://stackoverflow.com/a/45445291 for some retryable utility functions. I would prefer this approche as we could test each component individually (the retryable, the timeout, and finally the health state would be trivial)

Moreover, fetch is pretty bad at providing some custom timeout, we might not want wait for the maximum configured, see https://stackoverflow.com/questions/46946380/fetch-api-request-timeout

Comment (jeffmaury) [reply packages/backend/src/managers/playground.ts:210]: Delayed to a follow up PR

**Pull request merged**